### PR TITLE
chore(gateway): debug logging for the http requests

### PR DESCRIPTION
**Pull request by manute**
This PR tries to solve this issue https://github.com/ipfs/go-ipfs/issues/7873 

Please let me know if my approach is ok with the code and the issue.

Thanks!

### CODE REVIEW

**Review by lidel — CHANGES_REQUESTED** [review]
Thank you @manute, this is useful!

Noticed some cosmetic nits to address before we can merge this. Listed them below.


## TODO

- [ ] document `ipfs log level core/server debug` in `/docs/gateway.md`
-  [ ] those `error (type nil)` messages are noisy, are really hard to parse. 
   To illustrate, `curl -H "Host: ipfs.io" 127.0.0.1:8080 ` produces:
   ```
   2021-10-29T22:08:13.646+0200	DEBUG	core/server	corehttp/gateway_handler.go:202	http request received	{"from": "/ipns/ipfs.io/"}
   2021-10-29T22:08:13.647+0200	DEBUG	core/server	corehttp/gateway_handler.go:390	error (type nil) getting file tree for index.html, serve the file	{"from": "/ipns/ipfs.io/", "path": "/ipfs/QmPgJLQHLi5fNLTC2DJABJTa9T7dHy55TvKzatrvQS54B2/index.html"}
   ```
   A simple `serving index.html file`  would be easier to reason about here.
- [ ] `localhost:8080/ipfs/bafybeic7etg7wgzgyhvkyl62iaa7klhxlatjw5uowccvzxxa72aeo5eqsu` → subdomain is not logged at all
- subdomain requests are logged with duplicated CID (once in host, once in path). Opening `http://bafybeic7etg7wgzgyhvkyl62iaa7klhxlatjw5uowccvzxxa72aeo5eqsu.ipfs.localhost:8080/`
produces:
  ```
  2021-10-29T21:57:17.142+0200	DEBUG	core/server	corehttp/gateway_handler.go:202	http request received	{"from": "http://bafybeic7etg7wgzgyhvkyl62iaa7klhxlatjw5uowccvzxxa72aeo5eqsu.ipfs.localhost:8080/ipfs/bafybeic7etg7wgzgyhvkyl62iaa7klhxlatjw5uowccvzxxa72aeo5eqsu/"}
  ```

We will have limited availability during the next week and a half, so if you have time to address the above during that time, would make it easier for us to include this in v0.11.0-rc1

**Comment by manute** [comment]
Hey @lidel, 

Thanks for the review, I have addressed some changes based on that. 
A few questions: 
> redirect from localhost:8080/ipfs/bafybeic7etg7wgzgyhvkyl62iaa7klhxlatjw5uowccvzxxa72aeo5eqsu to the subdomain is not logged at all

Where I can find those subdomain redirects? 


>  subdomain requests are logged with duplicated CID (once in host, once in path)

Do you have any ideas to remove that duplication? Shall I do that before this [line](https://github.com/ipfs/go-ipfs/pull/8518/files#diff-427f4a268b11d614e7c3b3fdf35040496dd19492e325e3b1bd4bdaa4f8bf8e66R201)? And do you have any utility library that parses the path w/o that duplication?

**Comment by lanzafame** [comment]
@manute @Stebalien It may be even better to create a separate logger from the `core/server` logger that logs the requests, i.e. an access log similar to nginx. That way one can just monitor gateway requests without getting internal server errors. And having it as a separate logger will make it easier to filter out with tools such as loki or elasticsearch.

**Comment by Stebalien** [comment]
> @manute @Stebalien It may be even better to create a separate logger from the core/server logger that logs the requests, i.e. an access log similar to nginx. That way one can just monitor gateway requests without getting internal server errors. And having it as a separate logger will make it easier to filter out with tools such as loki or elasticsearch.

That's a good point. Logging to `gateway/requests`, or something like that, would be nice.

> Where I can find those subdomain redirects?

https://github.com/ipfs/go-ipfs/blob/52a747763f6c4e85b33ca051cda9cc4b75c815f9/core/corehttp/hostname.go#L97

Ish. It's an interesting redirect because we return the new path, but _also_ process the request.

Maybe we should just log response information in https://github.com/ipfs/go-ipfs/blob/52a747763f6c4e85b33ca051cda9cc4b75c815f9/core/corehttp/gateway_handler.go#L83? That's probably the best approach and that might just cover everything.

> Do you have any ideas to remove that duplication? Shall I do that before this line? And do you have any utility library that parses the path w/o that duplication?

Hm. Maybe use the `RequestURI` instead? That's the URL that the user _requested_, before modifications, cleaning, etc.

**Comment by manute** [comment]
Hey @Stebalien @lidel , 
I think I've addressed all your feedback. Please, let me know what do you think. 

@lanzafame  re: 
> @manute @Stebalien It may be even better to create a separate logger from the core/server logger that logs the requests, i.e. an access log similar to nginx. That way one can just monitor gateway requests without getting internal server errors. And having it as a separate logger will make it easier to filter out with tools such as loki or elasticsearch.

I think it's a good idea although I'd prefer to do that in another PR along with another issue, as it could have other implications. 
 
Thanks!

File: core/corehttp/gateway_handler.go [hunk 1]
```diff
@@ -83,6 +83,7 @@ func (sw *statusResponseWriter) WriteHeader(code int) {
 	if redirect != "" && code == http.StatusOK {
 		code = http.StatusMovedPermanently
 	}
+	log.Debugw("subdomain redirect", "location", redirect, "status", code)
```
Comment (Stebalien) [inline core/corehttp/gateway_handler.go:86]: This will always log a redirect, even if we didn't redirect.

Comment (manute) [reply core/corehttp/gateway_handler.go]: 🤦 fixed, thanks!

**Comment by manute** [comment]
@Stebalien @lidel Hey , does it need more changes or is it all fine?

**Comment by hyperknot** [comment]
Can this be merged in? Otherwise I see no reason to debug gateway configuration issues.

**Review by guseggert — APPROVED** [review]
